acqua-frontend/src/app/components/tickets/registrar-pago/registrar-pago.component.ts

```diff
@@ -36,28 +36,119 @@ export class RegistrarPagoComponent
     this.ticket = ticket;
-    this.formPagos = this.fb.group({
-      monto: ['', [Validators.required, Validators.min(1), Validators.max(this.ticket.restante ?? 0)]],
+    this.formPagos = this.getFormValidationForCashPaymentType();
+    this.SubscribeToFormChangesInCashPaymentType();
+  }
+
+  private getFormValidationForCashPaymentType()
+  {
+    return this.fb.group({
+      monto: ['', [
+        Validators.required,
+        Validators.min(1),
+        Validators.max(this.ticket.restante ?? 1),
+        Validators.pattern(/^[0-9]+$/)],
+      ],
+      cantidadRecibida: ['', [
+        Validators.required,
+        Validators.min(1),
+        Validators.pattern(/^[0-9]+$/),
+      ]],
+      cantidadDevuelta: ['', [
+        Validators.required,
+        Validators.min(0),
+        Validators.pattern(/^[0-9]+$/),
+      ]],
       referencia: [''],
     });
   }
 
-  reImprimirTickets()
+  private getFormValidationForCardPaymentType()
   {
-    // eslint-disable-next-line max-len
-    // TODO: Renderizar el componente de ticket preview, agregando los botones para imprimir ambos tickets
+    return this.fb.group({
+      monto: [''],
+      cantidadRecibida: [''],
+      cantidadDevuelta: [''],
+      referencia: ['', [
+        Validators.required,
+        Validators.minLength(10),
+        Validators.maxLength(25),
+        Validators.pattern(/^[a-zA-Z0-9]+$/),
+      ]],
+    });
+  }
+
+  private SubscribeToFormChangesInCashPaymentType()
+  {
+    this.formPagos.controls['monto'].valueChanges.subscribe(() =>
+    {
+      this.ActualizarDevolucionCambio();
+    });
+    this.formPagos.controls['cantidadRecibida'].valueChanges.subscribe(() =>
+    {
+      this.ActualizarDevolucionCambio();
+    });
   }
 
   get monto()
   {
     return this.formPagos.controls['monto'];
   }
 
+  get cantidadRecibida()
+  {
+    return this.formPagos.controls['cantidadRecibida'];
+  }
+
+  get cantidadDevuelta()
+  {
+    return this.formPagos.controls['cantidadDevuelta'];
+  }
+
   get referencia()
   {
     return this.formPagos.controls['referencia'];
   }
 
+  ActualizarDevolucionCambio()
+  {
+    const anticipo_cliente = this.monto.value;
+    const cantidad_recibida = this.cantidadRecibida.value;
+    const devolucion_cambio = this.calcularDevolucionCambio(anticipo_cliente, cantidad_recibida);
+    this.formPagos.controls['cantidadDevuelta'].setValue(devolucion_cambio);
+  }
+
+  calcularDevolucionCambio(anticipo_cliente = 0, cantidad_recibida = 0): number | null
+  {
+    if(anticipo_cliente > 0 && cantidad_recibida > 0)
+    {
+      const devolucionCambio = cantidad_recibida - anticipo_cliente;
+      if(devolucionCambio >= 0)
+      {
+        return devolucionCambio;
+      }
+      return null;
+    }
+    else
+    {
+      return null;
+    }
+  }
+
+  changeMetodoPago(metodoPago: string)
+  {
+    this.metodoPago = metodoPago as MetodoPago;
+    if(this.metodoPago == MetodoPago.Tarjeta || this.metodoPago == MetodoPago.Transferencia)
```

Comment:
si se recibe metodoPago, por qué se hace referencia mediante `this`?

Reply:
Ya se corrigió. Aun que sigue modificando el estado de una variable de la clase.